### PR TITLE
[WIP] Add implementation of separate file share controller and dock processes

### PR DESCRIPTION
**Pull request by Shruthi-1MN**
**What this PR does / why we need it**:
Implementation of new separate controller and dock for file share feature - only post/create, delete file share has been done yet. Still driver code implementation is in progress. Integration test has to be done yet with driver code.

File share post code workflow is as below:
Controller workflow:
   1 gRPC interface(model/fileshareproto, filesharecontroller.go)
   2 Fileshare controller client(filesharecontroller/client)
   3 selector to select the profiles
   4 Filesharecontroller server(filesharecontroller/controller.go)
Dock workflow
   1 It checks with fileshareproto to validate the dock creation
   2 Client communicates to dock server(filesharedock/client)
   3 Server dock.go - which registers to the dock and provision the dock so it contacts dock/discovery to discover and maps to backend drivers

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/model/fileshareproto/model.proto [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    // Delete a file share
+    rpc DeleteFileShare (DeleteFileShareOpts) returns (GenericResponse){} 
+
+}
+
+// CreateFileShareOpts is a structure which indicates all required properties for creating a file share.
+message CreateFileShareOpts {
+    // The uuid of the file share, optional when creating.
+    string id = 1;
```
Comment (jackhaibo) [inline pkg/model/fileshareproto/model.proto]: `CreateFileShareOpts `lacks `shareProto` filed which represents `CIFS` or `NFS `file share protocol. `Oceanstor` file driver needs this filed. If there is a multi-tenant scenario, please add `tenantID `or `userID `to CreateFileShareOpts.

Comment (Shruthi-1MN) [reply pkg/model/fileshareproto/model.proto]: Access protocol information you can get from profile, please specify profile id in request which has accessprotocol name when you are creating fileshare. In request of file share create /v1beta/tenantid/file/shares, from api request you can get tenantid is a request parameter.

**Comment by Shruthi-1MN** [comment]
> Do not push binary files (build/out/..) to github

Removed

File: contrib/connector/nvmeof/nvmeof_helper.go [hunk 2]
```diff
@@ -147,7 +147,7 @@ func Connect(connMap map[string]interface{}) (string, error) {
 	targetPortal := conn.TgtPortal
 	port := conn.TgtPort
 	nvmeTransportType := "rdma"
-	log.Printf("conn information:%s, %s, %s ", connNqn, targetPortal,  port)
+	log.Printf("conn information:%s, %s, %s ", connNqn, targetPortal, port)
```
Comment (NajmudheenCT) [inline contrib/connector/nvmeof/nvmeof_helper.go]: These changes are not related File share. Are these changes related to goimports ? then probably you can ignore these files as Lean has a PR https://github.com/opensds/opensds/pull/718 for these.

File: contrib/drivers/filesharedrivers/drivers.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+type FileShareDriver interface {
+	//Any initialization the fileshare driver does while starting.
+	Setup() error
+	//Any operation the fileshare driver does while stopping.
+	Unset() error
+
+	CreateFileShare(opt *pb.CreateFileShareOpts) (*model.FileShareSpec, error)
+	//ListPools() (interface{}, interface{})
```
Comment (NajmudheenCT) [inline contrib/drivers/filesharedrivers/drivers.go]: Remove unwanted code

File: contrib/drivers/filesharedrivers/nfsnative/cli.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+func sizeStr(size int64) string {
+	return fmt.Sprintf("%dg", size)
+}
+
+func (c *Cli) Mount(lvPath string) error {
+	cmd := []string{
+		"env", "LC_ALL=C",
+		"mount",
+		lvPath,
+		"/var/pravin_repo3",
```
Comment (NajmudheenCT) [inline contrib/drivers/filesharedrivers/nfsnative/cli.go]: /var/pravin_repo3 What is this ? This has to be a passed string, is it?

File: contrib/drivers/filesharedrivers/nfsnative/cli.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	_, err := c.execute(cmd...)
+	return err
+}
+
+func (c *Cli) CreateDirectory() error {
+	cmd := []string{
+		"env", "LC_ALL=C",
+		"mkdir",
+		"/var/pravin_repo3",
```
Comment (NajmudheenCT) [inline contrib/drivers/filesharedrivers/nfsnative/cli.go]: same here

File: contrib/drivers/filesharedrivers/nfsnative/nfsnative.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 The OpenSDS Authors. All Rights Reserved.
```
Comment (NajmudheenCT) [inline contrib/drivers/filesharedrivers/nfsnative/nfsnative.go]: Remove 'All Rights Reserved'.

File: contrib/drivers/filesharedrivers/nfsnative/nfsnative.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+type Driver struct {
+	conf *NFSNativeConfig
+	cli  *Cli
+}
+
+func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
+	// TODO Sample test pool will load to DB for driver code test
+	var pols []*model.StoragePoolSpec
+	pols = append(pols, &SamplePools[1])
```
Comment (NajmudheenCT) [inline contrib/drivers/filesharedrivers/nfsnative/nfsnative.go]: What is &SamplePools , I could not see declaration of this.

File: contrib/drivers/filesharedrivers/nfsnative/nfsnative.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	return pols, nil
+}
+
+func (d *Driver) Setup() error {
+	d.conf = &NFSNativeConfig{ConfigFile: "/etc/nfsnative/nfsnative.conf"}
```
Comment (NajmudheenCT) [inline contrib/drivers/filesharedrivers/nfsnative/nfsnative.go]: make config path as a const

File: pkg/api/util/util/db_test.go [hunk 9]
```diff
@@ -0,0 +1,330 @@
+// Copyright (c) 2018 Huawei Technologies Co., Ltd. All Rights Reserved.
```
Comment (NajmudheenCT) [inline pkg/api/util/util/db_test.go]: Why is this file under /util/util . There is already /util/db_test.go. Hope this is copy accidentally created. please remove

File: pkg/filesharecontroller/controller.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	result.PoolId, result.ProfileId = opt.GetPoolId(), opt.GetProfileId()
+
+	// Update the file share data in database.
+	db.C.UpdateStatus(ctx, result, model.FileShareAvailable)
+
+	return pb.GenericResponseResult(result), nil
+}
+
+// DeleteVolume implements pb.ControllerServer.DeleteFileShare
```
Comment (NajmudheenCT) [inline pkg/filesharecontroller/controller.go]: change comment to DeleteFileshare

File: pkg/filesharecontroller/selector/filter.go [hunk 11]
```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2019 OpenSDS Authors. All Rights Reserved.
```
Comment (NajmudheenCT) [inline pkg/filesharecontroller/selector/filter.go]: Hope controller/selector/filter.go can be reused.

Comment (Shruthi-1MN) [reply pkg/filesharecontroller/selector/filter.go]: selector i can't but filter we can

File: pkg/filesharecontroller/controller.go (hunk 10 above)
Comment (Shruthi-1MN) [reply pkg/filesharecontroller/controller.go]: Yes i am doing

File: contrib/connector/nvmeof/nvmeof_helper.go (hunk 2 above)
Comment (Shruthi-1MN) [reply contrib/connector/nvmeof/nvmeof_helper.go]: mistake, need to remove

Comment (Shruthi-1MN) [reply contrib/connector/nvmeof/nvmeof_helper.go]: Need to run go fmt then everything will be fine

File: contrib/drivers/filesharedrivers/nfsnative/nfsnative.go (hunk 7 above)
Comment (Shruthi-1MN) [reply contrib/drivers/filesharedrivers/nfsnative/nfsnative.go]: it is dummy driver, it is for testing dock code

File: contrib/drivers/filesharedrivers/nfsnative/nfsnative.go (hunk 8 above)
Comment (Shruthi-1MN) [reply contrib/drivers/filesharedrivers/nfsnative/nfsnative.go]: ok

**Comment by skdwriting** [comment]
@Shruthi-1MN , Is this still WIP? Can we remove if it is a PR ready for review and merge?

**Comment by skdwriting** [comment]
@Shruthi-1MN , there are conflicts reported, please check

**Comment by skdwriting** [comment]
@Shruthi-1MN , too many files changes in single PR. Possible to split logically?

File: install/devsds/lib/lvm.sh [hunk 12]
```diff
@@ -194,10 +253,21 @@ osds::lvm::clean_backing_file() {
     fi
 }
 
+osds::lvm::clean_volume_group_for_file() {
+    local fvg="opensds-files-default"
+    osds::lvm::remove_volume_group_for_file $fvg
+    # if there is no logical volume left, it's safe to attempt a cleanup
+    # of the backing file
+    if [[ -z "$(sudo lvs --noheadings -o lv_name $fvg 2>/dev/null)" ]]; then
```
Comment (kumarashit) [inline install/devsds/lib/lvm.sh]: wondering why we are using lvs somewhere and lvm most places? can't we use lvm everywhere. Also, command absolute path should be used.

File: pkg/api/controllers/fileshare.go [hunk 13]
```diff
@@ -50,6 +54,26 @@ func (f *FileSharePortal) CreateFileShare() {
 		log.Error(reason)
 		return
 	}
+
+	// Get profile
+	var prf *model.ProfileSpec
+	var err error
+	if fileshare.ProfileId == "" {
```
Comment (kumarashit) [inline pkg/api/controllers/fileshare.go]: Is profileId being passed? If yes, then why the Else condition

File: cmd/osdsfiledock/osdsfiledock.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	logs.InitLogs(CONF.OsdsfileDock.LogFlushFrequency)
+	defer logs.FlushLogs()
+
+	// Set up database session.
+	db.Init(&CONF.Database)
+
+	// Construct dock module grpc server struct and run dock server process.
+	ds := filesharedock.NewDockServer(CONF.OsdsfileDock.DockType, CONF.OsdsfileDock.ApiEndpoint)
+	if err := ds.Run(); err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (Shruthi-1MN) [reply cmd/osdsfiledock/osdsfiledock.go]: Good question. Sure i will.

File: pkg/api/controllers/fileshare.go (hunk 13 above)
Comment (Shruthi-1MN) [reply pkg/api/controllers/fileshare.go]: If it is not passed it should use default profile that is the reason and there will be warning message log.Warning("Use default profile when user doesn't specify profile.")

**Comment by Shruthi-1MN** [comment]
https://github.com/opensds/opensds/pull/726
